iroh-gateway/src/core.rs

```diff
@@ -0,0 +1,412 @@
+use axum::{
+    body::{self, Body, BoxBody},
+    error_handling::HandleErrorLayer,
+    extract::{Extension, Path, Query},
+    http::{header::*, StatusCode},
+    response::IntoResponse,
+    routing::get,
+    BoxError, Router,
+};
+use serde::Deserialize;
+use std::{borrow::Cow, collections::HashMap, time::Duration};
+use tower::ServiceBuilder;
+
+use crate::{
+    client::{Client, Request},
+    config::Config,
+    constants::*,
+    error::GatewayError,
+    response::{GatewayResponse, ResponseFormat},
+};
+
+pub struct Core {
+    pub config: Config,
+    client: Client,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct GetParams {
+    format: Option<String>,
+    filename: Option<String>,
+    download: Option<String>,
+}
+
+impl Core {
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            client: Client::new(),
+        }
+    }
+
+    pub async fn serve(&self) {
+        let app = Router::new()
+            .route("/ipfs/:cid", get(get_ipfs))
+            .route("/ipfs/:cid/*cpath", get(get_ipfs))
+            .layer(Extension(self.config.clone()))
+            .layer(Extension(self.client.clone()))
+            .layer(
+                ServiceBuilder::new()
+                    // Handle errors from middleware
+                    .layer(HandleErrorLayer::new(middleware_error_handler))
+                    .load_shed()
+                    .concurrency_limit(1024)
+                    .timeout(Duration::from_secs(10))
+                    .into_inner(),
+            );
+        let addr = format!("0.0.0.0:{}", self.config.port);
+        axum::Server::bind(&addr.parse().unwrap())
+            .http1_preserve_header_case(true)
+            .http1_title_case_headers(true)
+            .serve(app.into_make_service())
+            .await
+            .unwrap();
+    }
+}
+
+async fn get_ipfs(
+    Extension(config): Extension<Config>,
+    Extension(client): Extension<Client>,
+    Path(params): Path<HashMap<String, String>>,
+    Query(query_params): Query<GetParams>,
+) -> Result<GatewayResponse, GatewayError> {
+    // parse path params
+    let cid = params.get("cid").unwrap();
+    let cpath = "".to_string();
+    let cpath = params.get("cpath").unwrap_or(&cpath);
+    let full_content_path = format!("/ipfs/{}{}", cid, cpath);
+
+    // parse query params
+    let format = query_params.format.unwrap_or("".to_string());
```

Comment:
looking at this code `format` is not optional, so instead of unwraping it here, you should return an error immediately instead of trying to parse it, if it is `None`.

Reply:
`""` is an option and defaults to the `Fs` format.

Reply:
then you should be able to write sth like this

```rust
let format = if let Some(format) = query_params.format {
 match ResponseFormat::try_from(format.as_str()) {
     Ok(format) => format,
     Err(err) => {
       let msg = format!("{}", err);
       return error(StatusCode::BAD_REQUEST, &msg);
    }
 }
} else {
  ResponseFormat::Fs
};
```